### ISSUE TITLE
UCX 1.9.0-rc2 cannot be compiled against MOFED 4.0 libs

### ISSUE THREAD

**Issue by mkre** [openucx/ucx#5630]
### Describe the bug
We want to compile UCX to be used on systems where MOFED 4.x is installed. To that end, we are compiling UCX against the libraries and headers from MOFED 4.0. While evaluating UCX 1.9.0-rc2, we came across the following error at compilation:
<details>
  <summary>Click to expand!</summary>

```
  CC       mlx5/libuct_ib_la-ib_mlx5.lo
  CC       mlx5/exp/libuct_ib_la-ib_mlx5_hw.lo
  CC       mlx5/exp/libuct_ib_la-ib_exp.lo
  CC       mlx5/exp/libuct_ib_la-ib_exp_md.lo
In file included from mlx5/ib_mlx5_log.h:10:0,
                 from mlx5/ib_mlx5_log.c:11:
mlx5/ib_mlx5.h: In function 'uct_ib_mlx5_md_buf_alloc':
mlx5/ib_mlx5.h:591:17: error: implicit declaration of function 'mlx5dv_devx_umem_reg' [-Werror=implicit-function-declaration]
     mem->mem  = mlx5dv_devx_umem_reg(md->super.dev.ibv_context, buf, size, 0);
                 ^~~~~~~~~~~~~~~~~~~~
mlx5/ib_mlx5.h:591:17: error: nested extern declaration of 'mlx5dv_devx_umem_reg' [-Werror=nested-externs]
mlx5/ib_mlx5.h:591:15: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
     mem->mem  = mlx5dv_devx_umem_reg(md->super.dev.ibv_context, buf, size, 0);
               ^
mlx5/ib_mlx5.h: In function 'uct_ib_mlx5_md_buf_free':
mlx5/ib_mlx5.h:620:5: error: implicit declaration of function 'mlx5dv_devx_umem_dereg'; did you mean 'uct_md_mem_dereg'? [-Werror=implicit-function-declaration]
     mlx5dv_devx_umem_dereg(mem->mem);
     ^~~~~~~~~~~~~~~~~~~~~~
     uct_md_mem_dereg
mlx5/ib_mlx5.h:620:5: error: nested extern declaration of 'mlx5dv_devx_umem_dereg' [-Werror=nested-externs]
  CC       rc/base/libuct_ib_la-rc_ep.lo
In file included from mlx5/ib_mlx5.c:11:0:
mlx5/ib_mlx5.h: In function 'uct_ib_mlx5_md_buf_alloc':
mlx5/ib_mlx5.h:591:17: error: implicit declaration of function 'mlx5dv_devx_umem_reg' [-Werror=implicit-function-declaration]
     mem->mem  = mlx5dv_devx_umem_reg(md->super.dev.ibv_context, buf, size, 0);
                 ^~~~~~~~~~~~~~~~~~~~
mlx5/ib_mlx5.h:591:17: error: nested extern declaration of 'mlx5dv_devx_umem_reg' [-Werror=nested-externs]
mlx5/ib_mlx5.h:591:15: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
     mem->mem  = mlx5dv_devx_umem_reg(md->super.dev.ibv_context, buf, size, 0);
               ^
mlx5/ib_mlx5.h: In function 'uct_ib_mlx5_md_buf_free':
mlx5/ib_mlx5.h:620:5: error: implicit declaration of function 'mlx5dv_devx_umem_dereg'; did you mean 'uct_md_mem_dereg'? [-Werror=implicit-function-declaration]
     mlx5dv_devx_umem_dereg(mem->mem);
     ^~~~~~~~~~~~~~~~~~~~~~
     uct_md_mem_dereg
mlx5/ib_mlx5.h:620:5: error: nested extern declaration of 'mlx5dv_devx_umem_dereg' [-Werror=nested-externs]
mlx5/ib_mlx5.c: In function 'uct_ib_mlx5_devx_uar_init':
mlx5/ib_mlx5.c:393:27: error: implicit declaration of function 'mlx5dv_devx_alloc_uar' [-Werror=implicit-function-declaration]
     uar->uar            = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context, 0);
                           ^~~~~~~~~~~~~~~~~~~~~
mlx5/ib_mlx5.c:393:27: error: nested extern declaration of 'mlx5dv_devx_alloc_uar' [-Werror=nested-externs]
mlx5/ib_mlx5.c:393:25: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
     uar->uar            = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context, 0);
                         ^
  CC       rc/base/libuct_ib_la-rc_iface.lo
mlx5/ib_mlx5.c:399:35: error: dereferencing pointer to incomplete type 'struct mlx5dv_devx_uar'
     uar->super.addr.ptr = uar->uar->reg_addr;
                                   ^~
mlx5/ib_mlx5.c: In function 'uct_ib_mlx5_devx_uar_cleanup':
mlx5/ib_mlx5.c:412:5: error: implicit declaration of function 'mlx5dv_devx_free_uar' [-Werror=implicit-function-declaration]
     mlx5dv_devx_free_uar(uar->uar);
     ^~~~~~~~~~~~~~~~~~~~
mlx5/ib_mlx5.c:412:5: error: nested extern declaration of 'mlx5dv_devx_free_uar' [-Werror=nested-externs]
mlx5/ib_mlx5.c: In function 'uct_ib_mlx5_verbs_srq_init':
mlx5/ib_mlx5.c:585:29: error: 'MLX5DV_SRQ_MASK_SRQN' undeclared (first use in this function); did you mean 'MLX5DV_OBJ_SRQ'?
     srq_info.dv.comp_mask = MLX5DV_SRQ_MASK_SRQN;
                             ^~~~~~~~~~~~~~~~~~~~
                             MLX5DV_OBJ_SRQ
mlx5/ib_mlx5.c:585:29: note: each undeclared identifier is reported only once for each function it appears in
mlx5/ib_mlx5.c:594:31: error: 'struct mlx5dv_srq' has no member named 'srqn'
     srq->srq_num = srq_info.dv.srqn;
                               ^
cc1: all warnings being treated as errors
gmake[4]: *** [mlx5/libuct_ib_la-ib_mlx5.lo] Error 1
gmake[4]: *** Waiting for unfinished jobs....
cc1: all warnings being treated as errors
gmake[4]: *** [mlx5/libuct_ib_la-ib_mlx5_log.lo] Error 1
In file included from mlx5/exp/ib_exp_md.c:11:0:
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h: In function 'uct_ib_mlx5_md_buf_alloc':
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:591:17: error: implicit declaration of function 'mlx5dv_devx_umem_reg' [-Werror=implicit-function-declaration]
     mem->mem  = mlx5dv_devx_umem_reg(md->super.dev.ibv_context, buf, size, 0);
                 ^~~~~~~~~~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:591:17: error: nested extern declaration of 'mlx5dv_devx_umem_reg' [-Werror=nested-externs]
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:591:15: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
     mem->mem  = mlx5dv_devx_umem_reg(md->super.dev.ibv_context, buf, size, 0);
               ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h: In function 'uct_ib_mlx5_md_buf_free':
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:620:5: error: implicit declaration of function 'mlx5dv_devx_umem_dereg'; did you mean 'uct_md_mem_dereg'? [-Werror=implicit-function-declaration]
     mlx5dv_devx_umem_dereg(mem->mem);
     ^~~~~~~~~~~~~~~~~~~~~~
     uct_md_mem_dereg
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:620:5: error: nested extern declaration of 'mlx5dv_devx_umem_dereg' [-Werror=nested-externs]
cc1: all warnings being treated as errors
```
</details>

With `--with-mlx5-dv=no`, the error is different:
<details>
  <summary>Click to expand!</summary>

```
  CC       ud/base/libuct_ib_la-ud_iface_common.lo
  CC       ud/base/libuct_ib_la-ud_iface.lo
  CC       ud/base/libuct_ib_la-ud_ep.lo
  CC       ud/base/libuct_ib_la-ud_log.lo
  CC       ud/verbs/libuct_ib_la-ud_verbs.lo
In file included from rc/accel/rc_mlx5_common.h:13:0,
                 from rc/accel/rc_mlx5.h:10,
                 from rc/accel/rc_mlx5.inl:7,
                 from rc/accel/rc_mlx5_devx.c:11:
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:93:8: error: redefinition of 'struct mlx5_grh_av'
 struct mlx5_grh_av {
        ^~~~~~~~~~~
In file included from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:38:0,
                 from rc/accel/rc_mlx5_common.h:13,
                 from rc/accel/rc_mlx5.h:10,
                 from rc/accel/rc_mlx5.inl:7,
                 from rc/accel/rc_mlx5_devx.c:11:
/u/ydfb4q/.gradle/caches/cda/extLib/mofed-4.0-linux-x86_64/usr/include/infiniband/mlx5_hw.h:208:8: note: originally defined here
 struct mlx5_grh_av {
        ^~~~~~~~~~~
In file included from rc/accel/rc_mlx5_common.h:13:0,
                 from rc/accel/rc_mlx5.h:10,
                 from rc/accel/rc_mlx5.inl:7,
                 from rc/accel/rc_mlx5_devx.c:11:
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h: In function 'uct_ib_mlx5_md_buf_alloc':
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:591:17: error: implicit declaration of function 'mlx5dv_devx_umem_reg' [-Werror=implicit-function-declaration]
     mem->mem  = mlx5dv_devx_umem_reg(md->super.dev.ibv_context, buf, size, 0);
                 ^~~~~~~~~~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:591:17: error: nested extern declaration of 'mlx5dv_devx_umem_reg' [-Werror=nested-externs]
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:591:15: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
     mem->mem  = mlx5dv_devx_umem_reg(md->super.dev.ibv_context, buf, size, 0);
               ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h: In function 'uct_ib_mlx5_md_buf_free':
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:620:5: error: implicit declaration of function 'mlx5dv_devx_umem_dereg'; did you mean 'uct_md_mem_dereg'? [-Werror=implicit-function-declaration]
     mlx5dv_devx_umem_dereg(mem->mem);
     ^~~~~~~~~~~~~~~~~~~~~~
     uct_md_mem_dereg
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:620:5: error: nested extern declaration of 'mlx5dv_devx_umem_dereg' [-Werror=nested-externs]
In file included from rc/accel/rc_mlx5.inl:11:0,
                 from rc/accel/rc_mlx5_devx.c:11:
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl: In function 'uct_ib_mlx5_set_dgram_seg':
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:256:31: error: 'struct mlx5_wqe_av' has no member named 'key'
         mlx5_av_base(&seg->av)->key.qkey.qkey  = htonl(UCT_IB_KEY);
                               ^~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:264:27: error: 'struct mlx5_wqe_av' has no member named 'dqp_dct'
     mlx5_av_base(&seg->av)->dqp_dct            = av->dqp_dct;
                           ^~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:265:27: error: 'struct mlx5_wqe_av' has no member named 'stat_rate_sl'
     mlx5_av_base(&seg->av)->stat_rate_sl       = av->stat_rate_sl;
                           ^~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:266:27: error: 'struct mlx5_wqe_av' has no member named 'fl_mlid'
     mlx5_av_base(&seg->av)->fl_mlid            = av->fl_mlid;
                           ^~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:267:27: error: 'struct mlx5_wqe_av' has no member named 'rlid'
     mlx5_av_base(&seg->av)->rlid               = av->rlid;
                           ^~
In file included from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/type/status.h:12:0,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_verbs.h:21,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_device.h:10,
                 from rc/accel/rc_mlx5_common.h:10,
                 from rc/accel/rc_mlx5.h:10,
                 from rc/accel/rc_mlx5.inl:7,
                 from rc/accel/rc_mlx5_devx.c:11:
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:125:35: error: 'struct mlx5_wqe_av' has no member named 'reserved0'
     ((unsigned long)&( ((_type*)0)->_member ))
                                   ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:87:38: note: in expansion of macro 'ucs_offsetof'
                                      ucs_offsetof(struct mlx5_wqe_av, reserved0[0])))
                                      ^~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:272:16: note: in expansion of macro 'mlx5_av_grh'
         memcpy(mlx5_av_grh(&seg->av)->rmac, grh_av->rmac,
                ^~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:125:35: error: 'struct mlx5_wqe_av' has no member named 'reserved0'
     ((unsigned long)&( ((_type*)0)->_member ))
                                   ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:87:38: note: in expansion of macro 'ucs_offsetof'
                                      ucs_offsetof(struct mlx5_wqe_av, reserved0[0])))
                                      ^~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:273:23: note: in expansion of macro 'mlx5_av_grh'
                sizeof(mlx5_av_grh(&seg->av)->rmac));
                       ^~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:125:35: error: 'struct mlx5_wqe_av' has no member named 'reserved0'
     ((unsigned long)&( ((_type*)0)->_member ))
                                   ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:87:38: note: in expansion of macro 'ucs_offsetof'
                                      ucs_offsetof(struct mlx5_wqe_av, reserved0[0])))
                                      ^~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:275:9: note: in expansion of macro 'mlx5_av_grh'
         mlx5_av_grh(&seg->av)->tclass      = grh_av->tclass;
         ^~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:125:35: error: 'struct mlx5_wqe_av' has no member named 'reserved0'
     ((unsigned long)&( ((_type*)0)->_member ))
                                   ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:87:38: note: in expansion of macro 'ucs_offsetof'
                                      ucs_offsetof(struct mlx5_wqe_av, reserved0[0])))
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:276:9: note: in expansion of macro 'mlx5_av_grh'
         mlx5_av_grh(&seg->av)->hop_limit   = grh_av->hop_limit;
         ^~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:125:35: error: 'struct mlx5_wqe_av' has no member named 'reserved0'
     ((unsigned long)&( ((_type*)0)->_member ))
                                   ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:87:38: note: in expansion of macro 'ucs_offsetof'
                                      ucs_offsetof(struct mlx5_wqe_av, reserved0[0])))
                                      ^~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:277:9: note: in expansion of macro 'mlx5_av_grh'
         mlx5_av_grh(&seg->av)->grh_gid_fl  = grh_av->grh_gid_fl;
         ^~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:125:35: error: 'struct mlx5_wqe_av' has no member named 'reserved0'
     ((unsigned long)&( ((_type*)0)->_member ))
                                   ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:87:38: note: in expansion of macro 'ucs_offsetof'
                                      ucs_offsetof(struct mlx5_wqe_av, reserved0[0])))
                                      ^~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:278:16: note: in expansion of macro 'mlx5_av_grh'
         memcpy(mlx5_av_grh(&seg->av)->rgid, grh_av->rgid,
                ^~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:125:35: error: 'struct mlx5_wqe_av' has no member named 'reserved0'
     ((unsigned long)&( ((_type*)0)->_member ))
                                   ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:87:38: note: in expansion of macro 'ucs_offsetof'
                                      ucs_offsetof(struct mlx5_wqe_av, reserved0[0])))
                                      ^~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:279:23: note: in expansion of macro 'mlx5_av_grh'
                sizeof(mlx5_av_grh(&seg->av)->rgid));
                       ^~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:125:35: error: 'struct mlx5_wqe_av' has no member named 'reserved0'
     ((unsigned long)&( ((_type*)0)->_member ))
                                   ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.h:87:38: note: in expansion of macro 'ucs_offsetof'
                                      ucs_offsetof(struct mlx5_wqe_av, reserved0[0])))
                                      ^~~~~~~~~~~~
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/ib_mlx5.inl:281:9: note: in expansion of macro 'mlx5_av_grh'
         mlx5_av_grh(&seg->av)->grh_gid_fl  = 0;
         ^~~~~~~~~~~
In file included from /usr/include/bits/byteswap.h:35:0,
                 from /usr/include/endian.h:60,
                 from /usr/include/pthread.h:22,
                 from /u/ydfb4q/.gradle/caches/cda/extLib/mofed-4.0-linux-x86_64/usr/include/infiniband/verbs.h:40,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_verbs.h:14,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_device.h:10,
                 from rc/accel/rc_mlx5_common.h:10,
                 from rc/accel/rc_mlx5.h:10,
                 from rc/accel/rc_mlx5.inl:7,
                 from rc/accel/rc_mlx5_devx.c:11:
rc/accel/rc_mlx5.inl: In function 'uct_rc_mlx5_iface_tag_consumed':
rc/accel/rc_mlx5.inl:1158:38: error: 'struct mlx5_cqe64' has no member named 'app_info'; did you mean 'vlan_info'?
     tag = &iface->tm.list[ntohs(cqe->app_info)];
                                      ^
rc/accel/rc_mlx5.inl: In function 'uct_rc_mlx5_iface_handle_expected':
rc/accel/rc_mlx5.inl:1182:44: error: 'struct mlx5_cqe64' has no member named 'app_info'; did you mean 'vlan_info'?
     tag_entry = &iface->tm.list[ntohs(cqe->app_info)];
                                            ^
In file included from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/type/status.h:12:0,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_verbs.h:21,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_device.h:10,
                 from rc/accel/rc_mlx5_common.h:10,
                 from rc/accel/rc_mlx5.h:10,
                 from rc/accel/rc_mlx5.inl:7,
                 from rc/accel/rc_mlx5_devx.c:11:
rc/accel/rc_mlx5.inl: In function 'uct_rc_mlx5_iface_common_poll_rx':
rc/accel/rc_mlx5.inl:1411:23: error: 'struct mlx5_cqe64' has no member named 'app_op'
     if (ucs_likely(cqe->app_op == UCT_RC_MLX5_CQE_APP_OP_TM_UNEXPECTED)) {
                       ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/sys/compiler_def.h:180:53: note: in definition of macro 'ucs_likely'
 #define ucs_likely(x)              __builtin_expect(x, 1)
                                                     ^
In file included from rc/accel/rc_mlx5_devx.c:11:0:
rc/accel/rc_mlx5.inl:1416:16: error: 'struct mlx5_cqe64' has no member named 'app_op'
     switch (cqe->app_op) {
                ^~
In file included from /usr/include/bits/byteswap.h:35:0,
                 from /usr/include/endian.h:60,
                 from /usr/include/pthread.h:22,
                 from /u/ydfb4q/.gradle/caches/cda/extLib/mofed-4.0-linux-x86_64/usr/include/infiniband/verbs.h:40,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_verbs.h:14,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_device.h:10,
                 from rc/accel/rc_mlx5_common.h:10,
                 from rc/accel/rc_mlx5.h:10,
                 from rc/accel/rc_mlx5.inl:7,
                 from rc/accel/rc_mlx5_devx.c:11:
rc/accel/rc_mlx5.inl:1469:43: error: 'struct mlx5_cqe64' has no member named 'app_info'; did you mean 'vlan_info'?
         tag  = &iface->tm.list[ntohs(cqe->app_info)];
                                           ^
In file included from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/datastruct/arbiter.h:14:0,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/base/uct_iface.h:15,
                 from /u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/base/ib_device.h:13,
                 from rc/accel/rc_mlx5_common.h:10,
                 from rc/accel/rc_mlx5.h:10,
                 from rc/accel/rc_mlx5.inl:7,
                 from rc/accel/rc_mlx5_devx.c:11:
rc/accel/rc_mlx5.inl:1476:55: error: 'struct mlx5_cqe64' has no member named 'app_op'                                                                                                                                                                                   [40/98823]
         ucs_fatal("Unsupported packet arrived %d", cqe->app_op);
                                                       ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/ucs/debug/assert.h:51:47: note: in definition of macro 'ucs_fatal'
                            "Fatal: " _fmt, ## __VA_ARGS__)
                                               ^~~~~~~~~~~
rc/accel/rc_mlx5_devx.c: At top level:
rc/accel/rc_mlx5_devx.c:140:46: error: 'struct mlx5dv_pd' declared inside parameter list will not be visible outside of this definition or declaration [-Werror]
                                 const struct mlx5dv_pd *dvpd, void *wq)
                                              ^~~~~~~~~
In file included from rc/accel/rc_mlx5_devx.c:17:0:
rc/accel/rc_mlx5_devx.c: In function 'uct_rc_mlx5_devx_init_rx_common':
rc/accel/rc_mlx5_devx.c:177:52: error: dereferencing pointer to incomplete type 'const struct mlx5dv_pd'
     UCT_IB_MLX5DV_SET  (wq, wq, pd,            dvpd->pdn);
                                                    ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:39:25: note: in definition of macro 'UCT_IB_MLX5DV_SET'
         uint32_t ___v = _v; \
                         ^~
rc/accel/rc_mlx5_devx.c:180:74: error: dereferencing pointer to incomplete type 'struct mlx5dv_devx_umem'
     UCT_IB_MLX5DV_SET  (wq, wq, wq_umem_id,    iface->rx.srq.devx.mem.mem->umem_id);
                                                                          ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:39:25: note: in definition of macro 'UCT_IB_MLX5DV_SET'
         uint32_t ___v = _v; \
                         ^~
rc/accel/rc_mlx5_devx.c: In function 'uct_rc_mlx5_devx_init_rx_tm':
rc/accel/rc_mlx5_devx.c:214:12: error: variable 'dvpd' has initializer but incomplete type
     struct mlx5dv_pd dvpd = {};
            ^~~~~~~~~
rc/accel/rc_mlx5_devx.c:214:22: error: storage size of 'dvpd' isn't known
     struct mlx5dv_pd dvpd = {};
                      ^~~~
rc/accel/rc_mlx5_devx.c:224:7: error: 'struct mlx5dv_obj' has no member named 'pd'
     dv.pd.in  = uct_ib_iface_md(&iface->super.super)->pd;
       ^
rc/accel/rc_mlx5_devx.c:226:7: error: 'struct mlx5dv_obj' has no member named 'pd'
     dv.pd.out = &dvpd;
       ^
rc/accel/rc_mlx5_devx.c:228:5: error: implicit declaration of function 'mlx5dv_init_obj'; did you mean 'uct_ib_mlx5dv_init_obj'? [-Werror=implicit-function-declaration]
     mlx5dv_init_obj(&dv, MLX5DV_OBJ_PD | MLX5DV_OBJ_CQ);
     ^~~~~~~~~~~~~~~
     uct_ib_mlx5dv_init_obj
rc/accel/rc_mlx5_devx.c:228:5: error: nested extern declaration of 'mlx5dv_init_obj' [-Werror=nested-externs]
rc/accel/rc_mlx5_devx.c:228:26: error: 'MLX5DV_OBJ_PD' undeclared (first use in this function); did you mean 'MLX5DV_OBJ_CQ'?
     mlx5dv_init_obj(&dv, MLX5DV_OBJ_PD | MLX5DV_OBJ_CQ);
                          ^~~~~~~~~~~~~
                          MLX5DV_OBJ_CQ
rc/accel/rc_mlx5_devx.c:228:26: note: each undeclared identifier is reported only once for each function it appears in
rc/accel/rc_mlx5_devx.c:246:30: error: implicit declaration of function 'mlx5dv_devx_obj_create' [-Werror=implicit-function-declaration]
     iface->rx.srq.devx.obj = mlx5dv_devx_obj_create(dev->ibv_context,
                              ^~~~~~~~~~~~~~~~~~~~~~
rc/accel/rc_mlx5_devx.c:246:30: error: nested extern declaration of 'mlx5dv_devx_obj_create' [-Werror=nested-externs]
rc/accel/rc_mlx5_devx.c:246:28: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
     iface->rx.srq.devx.obj = mlx5dv_devx_obj_create(dev->ibv_context,
                            ^
rc/accel/rc_mlx5_devx.c:214:22: error: unused variable 'dvpd' [-Werror=unused-variable]
     struct mlx5dv_pd dvpd = {};
                      ^~~~
rc/accel/rc_mlx5_devx.c: In function 'uct_rc_mlx5_devx_init_rx':
rc/accel/rc_mlx5_devx.c:272:12: error: variable 'dvpd' has initializer but incomplete type
     struct mlx5dv_pd dvpd = {};
            ^~~~~~~~~
rc/accel/rc_mlx5_devx.c:272:22: error: storage size of 'dvpd' isn't known
     struct mlx5dv_pd dvpd = {};
                      ^~~~
rc/accel/rc_mlx5_devx.c:279:7: error: 'struct mlx5dv_obj' has no member named 'pd'
     dv.pd.in  = uct_ib_iface_md(&iface->super.super)->pd;
       ^
rc/accel/rc_mlx5_devx.c:280:7: error: 'struct mlx5dv_obj' has no member named 'pd'
     dv.pd.out = &dvpd;
       ^
rc/accel/rc_mlx5_devx.c:281:26: error: 'MLX5DV_OBJ_PD' undeclared (first use in this function); did you mean 'MLX5DV_OBJ_CQ'?
     mlx5dv_init_obj(&dv, MLX5DV_OBJ_PD);
                          ^~~~~~~~~~~~~
                          MLX5DV_OBJ_CQ
rc/accel/rc_mlx5_devx.c:294:28: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
     iface->rx.srq.devx.obj = mlx5dv_devx_obj_create(dev->ibv_context,
                            ^
rc/accel/rc_mlx5_devx.c:272:22: error: unused variable 'dvpd' [-Werror=unused-variable]
     struct mlx5dv_pd dvpd = {};
                      ^~~~
rc/accel/rc_mlx5_devx.c: In function 'uct_rc_mlx5_iface_common_devx_connect_qp':
rc/accel/rc_mlx5_devx.c:354:24: error: 'struct mlx5_wqe_av' has no member named 'rmac'
                &mlx5_av.rmac, sizeof(mlx5_av.rmac));
                        ^
rc/accel/rc_mlx5_devx.c:354:45: error: 'struct mlx5_wqe_av' has no member named 'rmac'
                &mlx5_av.rmac, sizeof(mlx5_av.rmac));
                                             ^
rc/accel/rc_mlx5_devx.c:356:24: error: 'struct mlx5_wqe_av' has no member named 'rgid'
                &mlx5_av.rgid, sizeof(mlx5_av.rgid));
                        ^
rc/accel/rc_mlx5_devx.c:356:45: error: 'struct mlx5_wqe_av' has no member named 'rgid'
                &mlx5_av.rgid, sizeof(mlx5_av.rgid));
                                             ^
In file included from rc/accel/rc_mlx5_devx.c:17:0:
rc/accel/rc_mlx5_devx.c:358:34: error: 'struct mlx5_wqe_av' has no member named 'hop_limit'
                           mlx5_av.hop_limit);
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:39:25: note: in definition of macro 'UCT_IB_MLX5DV_SET'
         uint32_t ___v = _v; \
                         ^~
rc/accel/rc_mlx5_devx.c:391:38: error: 'true' undeclared (first use in this function); did you mean 'tee'?
     UCT_IB_MLX5DV_SET(qpc, qpc, rre, true);
                                      ^
/u/ydfb4q/ucx/build/1.9.0-rc2/src/src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:39:25: note: in definition of macro 'UCT_IB_MLX5DV_SET'
         uint32_t ___v = _v; \
                         ^~
cc1: all warnings being treated as errors
```
</details>

This issue did not appear with UCX 1.8.0. Also, everything is working fine when building UCX 1.9.0-rc2 against MOFED **5.0**.

Will it be possible to compile and use UCX 1.9.0 on MOFED 4 systems?

### Steps to Reproduce
- Compile UCX 1.9.0-rc2 against MOFED 4.0

### Setup and versions
- CentOS 7.8
- MLNX_OFED 4.0

### Additional information (depending on the issue)
[config.log](https://github.com/openucx/ucx/files/5142242/config.log)
[config-mlx5-dv-no.log](https://github.com/openucx/ucx/files/5142243/config-mlx5-dv-no.log)

**Comment by Artemy-Mellanox** [comment]
@mkre may it be that your setup have some leftovers from previous installations?
configure detects symbol MLX5DV_CONTEXT_FLAGS_DEVX under /usr/include or under some xxx_INCLUDE_PATH environment variable
could you plz serach mlx5dv.h files on the setup you running configure

**Comment by mkre** [comment]
@Artemy-Mellanox, you are right! There was a stray mlx5dv.h in `/usr/include/infiniband` which prevented correct cross compilation. I have removed the `rdma-core-devel` package from my build machine and it is now working fine. The UCX 1.8->1.9 upgrade coincided with a software update on the build machine (including an update of `rdma-core-devel`), which caused this issue, so I got confused about the root cause.

Thanks for the quick help!
Moritz
